supervisord/datadog_checks/supervisord/supervisord.py

```diff
@@ -122,32 +122,63 @@ def check(self, instance):
 
+        proc_regex_exclude = instance.get('proc_regex_exclude', [])
+        if not isinstance(proc_regex_exclude, list):
+            raise Exception("'proc_regex_exclude' should be a list of strings. e.g. %s" % [proc_regex_exclude])
+
         proc_names = instance.get('proc_names', [])
         if not isinstance(proc_names, list):
             raise Exception("'proc_names' should be a list of strings. e.g. %s" % [proc_names])
 
+        proc_names_exclude = instance.get('proc_names_exclude', [])
+        if not isinstance(proc_names_exclude, list):
+            raise Exception("'proc_names_exclude' should be a list of strings. e.g. %s" % [proc_names_exclude])
+
+        # Compile status mapping
+        status_mapping = DD_STATUS
+
+        status_mapping_override = instance.get('status_mapping_override', {})
+        if not isinstance(status_mapping_override, dict):
+            raise Exception("'status_mapping_override' should be a dictionary. e.g. %s" % [status_mapping_override])
+
+        if len(status_mapping_override) != 0:
+            reverse_process_status = {v: k for k, v in PROCESS_STATUS.items()}
+            for status, ddstatus in status_mapping_override.items():
+                if ddstatus in PROCESS_STATUS.values():
+                    status_mapping[status] = reverse_process_status[ddstatus]
+                else:
+                    raise Exception(
+                        "'status_mapping_override' should be a status mapping e.g. %s => %s" % [status, ddstatus]
+                    )
+
         # Collect information on each monitored process
         monitored_processes = []
 
         # monitor all processes if no filters were specified
         if len(proc_regex) == 0 and len(proc_names) == 0:
             monitored_processes = processes
-
-        for pattern, process in itertools.product(proc_regex, processes):
-            if re.match(pattern, process['name']) and process not in monitored_processes:
-                monitored_processes.append(process)
-
-        for process in processes:
-            if process['name'] in proc_names and process not in monitored_processes:
-                monitored_processes.append(process)
+        else:
+            for pattern, process in itertools.product(proc_regex, processes):
+                if re.match(pattern, process['name']) and process not in monitored_processes:
+                    monitored_processes.append(process)
+            for process in processes:
+                if process['name'] in proc_names and process not in monitored_processes:
+                    monitored_processes.append(process)
+
+        for pattern, process in itertools.product(proc_regex_exclude, monitored_processes):
+            if re.match(pattern, process['name']):
+                monitored_processes.remove(process)
+        for process in monitored_processes:
+            if process['name'] in proc_names_exclude:
+                monitored_processes.remove(process)
```

Comment:
```suggestion        
        monitored_processes = set()

        for process in processes:
            name = process['name']
            include = (
                any(re.search(rgx, name) for rgx in proc_regex) if proc_regex else True
                or (name in proc_names) if proc_names else True
            )
            if not include:
                continue # No need to check exclusions if process doesn't match inclusions.
            exclude = (
                any(re.search(rgx, name) for rgx in proc_regex_exclude) if proc_regex_exclude else False
                or (name in proc_names_exclude) if proc_names_exclude else False
            )
            if not exclude:
                monitored_processes.add(name, process['statename'])
```
Here's a suggestion to stream-line the logic a bit. A couple of ideas:

1. Use a `set` of monitored processes to ensure no duplicates. 
2. Only add a process to the set if it passes both inclusion and exclusion filters.

Note that we will have to change how we loop over `monitored_processes`:

```python
for proc_name, status in monitored_processes:
    ...
```

Reply:
Your suggestion makes it easier to read. :)
However, I don't think it's good for the `set` usage, as we need all the `process` items afterwards.
So I implemented your modifications without the `set`.